<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="use_sim_time" default="false"/>
  <arg name="debug" default="false"/>
  <arg name="yolo_model" default="yolov8m.pt"/>
  <arg name="input_topic" default="/limo/depth_camera_link/image_raw"/>
  <arg name="depth_topic" default="/limo/depth_camera_link/depth/image_raw"/>
  <arg name="result_topic" default="/yolo_result"/>
  <arg name="result_image_topic" default="/yolo_image"/>
  <arg name="object_position_topic" default="/object_position"/>
  <arg name="object_position_world_topic" default="/object_position_world"/>
  <arg name="conf_thres" default="0.25"/>
  <arg name="iou_thres" default="0.45"/>
  <arg name="max_det" default="300"/>
  <arg name="tracker" default="bytetrack.yaml"/>
  <arg name="device" default=""/>
  <arg name="result_conf" default="true"/>
  <arg name="result_line_width" default="1"/>
  <arg name="result_font_size" default="1"/>
  <arg name="result_font" default="Arial.ttf"/>
  <arg name="result_labels" default="true"/>
  <arg name="result_boxes" default="true"/>
  <arg name="camera_frame_id" default="depth_camera_link"/>
  <arg name="target_frame" default="world"/>
  <arg name="use_fixed_frame" default="true"/>
  <arg name="tf_timeout" default="1.0"/>

  <set_parameter name="use_sim_time" value="$(var use_sim_time)"/>

  <!-- Run the tracker node for object detection -->
  <node pkg="ultralytics_ros" exec="tracker_node.py" output="screen">
    <param name="yolo_model" value="$(var yolo_model)"/>
    <param name="input_topic" value="$(var input_topic)"/>
    <param name="result_topic" value="$(var result_topic)"/>
    <param name="result_image_topic" value="$(var result_image_topic)"/>
    <param name="conf_thres" value="$(var conf_thres)"/>
    <param name="iou_thres" value="$(var iou_thres)"/>
    <param name="max_det" value="$(var max_det)"/>
    <param name="tracker" value="$(var tracker)"/>
    <param name="device" value="$(var device)"/>
    <param name="result_conf" value="$(var result_conf)"/>
    <param name="result_line_width" value="$(var result_line_width)"/>
    <param name="result_font_size" value="$(var result_font_size)"/>
    <param name="result_font" value="$(var result_font)"/>
    <param name="result_labels" value="$(var result_labels)"/>
    <param name="result_boxes" value="$(var result_boxes)"/>
  </node>

  <!-- Run the object localizer node to convert 2D detections to 3D positions -->
  <node pkg="ultralytics_ros" exec="object_localizer_node.py" output="screen">
    <param name="depth_image_topic" value="$(var depth_topic)"/>
    <param name="yolo_result_topic" value="$(var result_topic)"/>
    <param name="object_position_topic" value="$(var object_position_topic)"/>
    <param name="frame_id" value="$(var camera_frame_id)"/>
  </node>

  <!-- Run the transform node to convert positions from camera frame to world frame at origin (0,0,0) -->
  <node pkg="ultralytics_ros" exec="object_transform_node.py" output="screen">
    <param name="input_topic" value="$(var object_position_topic)"/>
    <param name="output_topic" value="$(var object_position_world_topic)"/>
    <param name="target_frame" value="$(var target_frame)"/>
    <param name="use_fixed_frame" value="$(var use_fixed_frame)"/>
    <param name="tf_timeout" value="$(var tf_timeout)"/>
  </node>

  <!-- Conditional debug visualization -->
  <node if="$(var debug)" pkg="image_view" exec="image_view" output="screen">
    <remap from="image" to="$(var result_image_topic)"/>
  </node>
</launch>